src/backend/distributed/commands/truncate.c

```diff
@@ -93,19 +93,15 @@ EnsurePartitionTableNotReplicatedForTruncate(TruncateStmt *truncateStatement)
 
 	foreach(relationCell, truncateStatement->relations)
 	{
-		RangeVar *relationRV = (RangeVar *) lfirst(relationCell);
-		Relation relation = heap_openrv(relationRV, NoLock);
-		Oid relationId = RelationGetRelid(relation);
+		RangeVar *rangeVar = (RangeVar *) lfirst(relationCell);
+		Oid relationId = RangeVarGetRelid(rangeVar, NoLock, false);
 
 		if (!IsDistributedTable(relationId))
 		{
-			heap_close(relation, NoLock);
 			continue;
 		}
 
 		EnsurePartitionTableNotReplicated(relationId);
```

Comment:
Unfortunately this is not behaving exactly the same. This function acquires `AccessShareLock`, and then if we do `TRUNCATE` from the MX workers at the same time, we get into distributed deadlocks:

```SQL

cat /tmp/f.sql 
TRUNCATE users_table;

pgbench -h localhost -p 9700 -f /tmp/f.sql -T 10 -P 5 postgres &
pgbench -h localhost -p 9701 -f /tmp/f.sql -T 10 -P 5 postgres &
pgbench -h localhost -p 9702 -f /tmp/f.sql -T 10 -P 5 postgres &

lient 0 aborted in command 0 (SQL) of script 0; ERROR:  canceling the transaction since it was involved in a distributed deadlock

client 0 aborted in command 0 (SQL) of script 0; ERROR:  canceling the transaction since it was involved in a distributed deadlock

```

I guess, with heap_close, we used to release all the locks on the relation.